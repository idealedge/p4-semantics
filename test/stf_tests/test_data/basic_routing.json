{
  "program" : "basic_routing.p4",
  "__meta__" : {
    "version" : [2, 7],
    "compiler" : "https://github.com/p4lang/p4c"
  },
  "header_types" : [
    {
      "name" : "scalars_0",
      "id" : 0,
      "fields" : []
    },
    {
      "name" : "ethernet_t",
      "id" : 1,
      "fields" : [
        ["dstAddr", 48, false],
        ["srcAddr", 48, false],
        ["etherType", 16, false]
      ]
    },
    {
      "name" : "ipv4_t",
      "id" : 2,
      "fields" : [
        ["version", 4, false],
        ["ihl", 4, false],
        ["diffserv", 8, false],
        ["totalLen", 16, false],
        ["identification", 16, false],
        ["flags", 3, false],
        ["fragOffset", 13, false],
        ["ttl", 8, false],
        ["protocol", 8, false],
        ["hdrChecksum", 16, false],
        ["srcAddr", 32, false],
        ["dstAddr", 32, false]
      ]
    },
    {
      "name" : "ingress_metadata_t",
      "id" : 3,
      "fields" : [
        ["vrf", 12, false],
        ["bd", 16, false],
        ["nexthop_index", 16, false],
        ["_padding", 4, false]
      ]
    },
    {
      "name" : "standard_metadata",
      "id" : 4,
      "fields" : [
        ["ingress_port", 9, false],
        ["egress_spec", 9, false],
        ["egress_port", 9, false],
        ["clone_spec", 32, false],
        ["instance_type", 32, false],
        ["drop", 1, false],
        ["recirculate_port", 16, false],
        ["packet_length", 32, false],
        ["enq_timestamp", 32, false],
        ["enq_qdepth", 19, false],
        ["deq_timedelta", 32, false],
        ["deq_qdepth", 19, false],
        ["ingress_global_timestamp", 48, false],
        ["lf_field_list", 32, false],
        ["mcast_grp", 16, false],
        ["resubmit_flag", 1, false],
        ["egress_rid", 16, false],
        ["checksum_error", 1, false],
        ["_padding_0", 4, false]
      ]
    }
  ],
  "headers" : [
    {
      "name" : "scalars",
      "id" : 0,
      "header_type" : "scalars_0",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "standard_metadata",
      "id" : 1,
      "header_type" : "standard_metadata",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "ethernet",
      "id" : 2,
      "header_type" : "ethernet_t",
      "metadata" : false,
      "pi_omit" : true
    },
    {
      "name" : "ipv4",
      "id" : 3,
      "header_type" : "ipv4_t",
      "metadata" : false,
      "pi_omit" : true
    },
    {
      "name" : "ingress_metadata",
      "id" : 4,
      "header_type" : "ingress_metadata_t",
      "metadata" : true,
      "pi_omit" : true
    }
  ],
  "header_stacks" : [],
  "header_union_types" : [],
  "header_unions" : [],
  "header_union_stacks" : [],
  "field_lists" : [],
  "errors" : [
    ["NoError", 1],
    ["PacketTooShort", 2],
    ["NoMatch", 3],
    ["StackOutOfBounds", 4],
    ["HeaderTooShort", 5],
    ["ParserTimeout", 6]
  ],
  "enums" : [],
  "parsers" : [
    {
      "name" : "parser",
      "id" : 0,
      "init_state" : "start",
      "parse_states" : [
        {
          "name" : "parse_ethernet",
          "id" : 0,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "ethernet"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "0x0800",
              "mask" : null,
              "next_state" : "parse_ipv4"
            },
            {
              "value" : "default",
              "mask" : null,
              "next_state" : null
            }
          ],
          "transition_key" : [
            {
              "type" : "field",
              "value" : ["ethernet", "etherType"]
            }
          ]
        },
        {
          "name" : "parse_ipv4",
          "id" : 1,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "ipv4"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : null
            }
          ],
          "transition_key" : []
        },
        {
          "name" : "start",
          "id" : 2,
          "parser_ops" : [],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : "parse_ethernet"
            }
          ],
          "transition_key" : []
        }
      ]
    }
  ],
  "deparsers" : [
    {
      "name" : "deparser",
      "id" : 0,
      "order" : ["ethernet", "ipv4"]
    }
  ],
  "meter_arrays" : [],
  "counter_arrays" : [],
  "register_arrays" : [],
  "calculations" : [
    {
      "name" : "calc",
      "id" : 0,
      "algo" : "csum16",
      "input" : [
        {
          "type" : "field",
          "value" : ["ipv4", "version"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "ihl"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "diffserv"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "totalLen"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "identification"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "flags"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "fragOffset"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "ttl"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "protocol"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "srcAddr"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "dstAddr"]
        }
      ]
    },
    {
      "name" : "calc_0",
      "id" : 1,
      "algo" : "csum16",
      "input" : [
        {
          "type" : "field",
          "value" : ["ipv4", "version"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "ihl"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "diffserv"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "totalLen"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "identification"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "flags"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "fragOffset"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "ttl"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "protocol"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "srcAddr"]
        },
        {
          "type" : "field",
          "value" : ["ipv4", "dstAddr"]
        }
      ]
    }
  ],
  "learn_lists" : [],
  "actions" : [
    {
      "name" : "NoAction",
      "id" : 0,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "on_miss",
      "id" : 1,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "rewrite_src_dst_mac",
      "id" : 2,
      "runtime_data" : [
        {
          "name" : "smac",
          "bitwidth" : 48
        },
        {
          "name" : "dmac",
          "bitwidth" : 48
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ethernet", "srcAddr"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 209,
            "column" : 27,
            "source_fragment" : "smac, dmac) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ethernet", "dstAddr"]
            },
            {
              "type" : "runtime_data",
              "value" : 1
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 209,
            "column" : 33,
            "source_fragment" : "dmac) { ..."
          }
        }
      ]
    },
    {
      "name" : "NoAction",
      "id" : 3,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "NoAction",
      "id" : 4,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "NoAction",
      "id" : 5,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "NoAction",
      "id" : 6,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "NoAction",
      "id" : 7,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "set_vrf",
      "id" : 8,
      "runtime_data" : [
        {
          "name" : "vrf",
          "bitwidth" : 12
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ingress_metadata", "vrf"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 131,
            "column" : 15,
            "source_fragment" : "vrf) { ..."
          }
        }
      ]
    },
    {
      "name" : "on_miss",
      "id" : 9,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "on_miss",
      "id" : 10,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "on_miss",
      "id" : 11,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "fib_hit_nexthop",
      "id" : 12,
      "runtime_data" : [
        {
          "name" : "nexthop_index",
          "bitwidth" : 16
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ingress_metadata", "nexthop_index"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 145,
            "column" : 23,
            "source_fragment" : "nexthop_index) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ipv4", "ttl"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "&",
                  "left" : {
                    "type" : "expression",
                    "value" : {
                      "op" : "+",
                      "left" : {
                        "type" : "field",
                        "value" : ["ipv4", "ttl"]
                      },
                      "right" : {
                        "type" : "hexstr",
                        "value" : "0xff"
                      }
                    }
                  },
                  "right" : {
                    "type" : "hexstr",
                    "value" : "0xff"
                  }
                }
              }
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 147,
            "column" : 4,
            "source_fragment" : "subtract_from_field(ipv4.ttl, 1)"
          }
        }
      ]
    },
    {
      "name" : "fib_hit_nexthop",
      "id" : 13,
      "runtime_data" : [
        {
          "name" : "nexthop_index",
          "bitwidth" : 16
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ingress_metadata", "nexthop_index"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 145,
            "column" : 23,
            "source_fragment" : "nexthop_index) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ipv4", "ttl"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "&",
                  "left" : {
                    "type" : "expression",
                    "value" : {
                      "op" : "+",
                      "left" : {
                        "type" : "field",
                        "value" : ["ipv4", "ttl"]
                      },
                      "right" : {
                        "type" : "hexstr",
                        "value" : "0xff"
                      }
                    }
                  },
                  "right" : {
                    "type" : "hexstr",
                    "value" : "0xff"
                  }
                }
              }
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 147,
            "column" : 4,
            "source_fragment" : "subtract_from_field(ipv4.ttl, 1)"
          }
        }
      ]
    },
    {
      "name" : "set_egress_details",
      "id" : 14,
      "runtime_data" : [
        {
          "name" : "egress_spec",
          "bitwidth" : 9
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["standard_metadata", "egress_spec"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 174,
            "column" : 26,
            "source_fragment" : "egress_spec) { ..."
          }
        }
      ]
    },
    {
      "name" : "set_bd",
      "id" : 15,
      "runtime_data" : [
        {
          "name" : "bd",
          "bitwidth" : 16
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["ingress_metadata", "bd"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 117,
            "column" : 14,
            "source_fragment" : "bd) { ..."
          }
        }
      ]
    }
  ],
  "pipelines" : [
    {
      "name" : "ingress",
      "id" : 0,
      "init_table" : "node_2",
      "tables" : [
        {
          "name" : "port_mapping",
          "id" : 0,
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 121,
            "column" : 0,
            "source_fragment" : "table port_mapping { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["standard_metadata", "ingress_port"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 32768,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [15, 7],
          "actions" : ["set_bd", "NoAction"],
          "base_default_next" : "bd",
          "next_tables" : {
            "set_bd" : "bd",
            "NoAction" : "bd"
          },
          "default_entry" : {
            "action_id" : 7,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "bd",
          "id" : 1,
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 135,
            "column" : 0,
            "source_fragment" : "table bd { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["ingress_metadata", "bd"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 65536,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [8, 3],
          "actions" : ["set_vrf", "NoAction"],
          "base_default_next" : "ipv4_fib",
          "next_tables" : {
            "set_vrf" : "ipv4_fib",
            "NoAction" : "ipv4_fib"
          },
          "default_entry" : {
            "action_id" : 3,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "ipv4_fib",
          "id" : 2,
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 150,
            "column" : 0,
            "source_fragment" : "table ipv4_fib { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["ingress_metadata", "vrf"],
              "mask" : null
            },
            {
              "match_type" : "exact",
              "target" : ["ipv4", "dstAddr"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 131072,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [9, 12, 4],
          "actions" : ["on_miss", "fib_hit_nexthop", "NoAction"],
          "base_default_next" : "nexthop",
          "next_tables" : {
            "on_miss" : "ipv4_fib_lpm",
            "fib_hit_nexthop" : "nexthop",
            "NoAction" : "nexthop"
          },
          "default_entry" : {
            "action_id" : 4,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "ipv4_fib_lpm",
          "id" : 3,
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 162,
            "column" : 0,
            "source_fragment" : "table ipv4_fib_lpm { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["ingress_metadata", "vrf"],
              "mask" : null
            },
            {
              "match_type" : "lpm",
              "target" : ["ipv4", "dstAddr"],
              "mask" : null
            }
          ],
          "match_type" : "lpm",
          "type" : "simple",
          "max_size" : 16384,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [10, 13, 5],
          "actions" : ["on_miss", "fib_hit_nexthop", "NoAction"],
          "base_default_next" : "nexthop",
          "next_tables" : {
            "on_miss" : "nexthop",
            "fib_hit_nexthop" : "nexthop",
            "NoAction" : "nexthop"
          },
          "default_entry" : {
            "action_id" : 5,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "nexthop",
          "id" : 4,
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 178,
            "column" : 0,
            "source_fragment" : "table nexthop { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["ingress_metadata", "nexthop_index"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 32768,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [11, 14, 6],
          "actions" : ["on_miss", "set_egress_details", "NoAction"],
          "base_default_next" : null,
          "next_tables" : {
            "on_miss" : null,
            "set_egress_details" : null,
            "NoAction" : null
          },
          "default_entry" : {
            "action_id" : 6,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        }
      ],
      "action_profiles" : [],
      "conditionals" : [
        {
          "name" : "node_2",
          "id" : 0,
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 190,
            "column" : 8,
            "source_fragment" : "valid(ipv4)"
          },
          "expression" : {
            "type" : "expression",
            "value" : {
              "op" : "==",
              "left" : {
                "type" : "field",
                "value" : ["ipv4", "$valid$"]
              },
              "right" : {
                "type" : "hexstr",
                "value" : "0x01"
              }
            }
          },
          "false_next" : null,
          "true_next" : "port_mapping"
        }
      ]
    },
    {
      "name" : "egress",
      "id" : 1,
      "init_table" : "rewrite_mac",
      "tables" : [
        {
          "name" : "rewrite_mac",
          "id" : 5,
          "source_info" : {
            "filename" : "basic_routing.p4",
            "line" : 214,
            "column" : 0,
            "source_fragment" : "table rewrite_mac { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["ingress_metadata", "nexthop_index"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 32768,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [1, 2, 0],
          "actions" : ["on_miss", "rewrite_src_dst_mac", "NoAction"],
          "base_default_next" : null,
          "next_tables" : {
            "on_miss" : null,
            "rewrite_src_dst_mac" : null,
            "NoAction" : null
          },
          "default_entry" : {
            "action_id" : 0,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        }
      ],
      "action_profiles" : [],
      "conditionals" : []
    }
  ],
  "checksums" : [
    {
      "name" : "cksum",
      "id" : 0,
      "target" : ["ipv4", "hdrChecksum"],
      "type" : "generic",
      "calculation" : "calc"
    },
    {
      "name" : "cksum_0",
      "id" : 1,
      "target" : ["ipv4", "hdrChecksum"],
      "type" : "generic",
      "calculation" : "calc_0"
    }
  ],
  "force_arith" : [],
  "extern_instances" : [],
  "field_aliases" : [
    [
      "queueing_metadata.enq_timestamp",
      ["standard_metadata", "enq_timestamp"]
    ],
    [
      "queueing_metadata.enq_qdepth",
      ["standard_metadata", "enq_qdepth"]
    ],
    [
      "queueing_metadata.deq_timedelta",
      ["standard_metadata", "deq_timedelta"]
    ],
    [
      "queueing_metadata.deq_qdepth",
      ["standard_metadata", "deq_qdepth"]
    ],
    [
      "intrinsic_metadata.ingress_global_timestamp",
      ["standard_metadata", "ingress_global_timestamp"]
    ],
    [
      "intrinsic_metadata.lf_field_list",
      ["standard_metadata", "lf_field_list"]
    ],
    [
      "intrinsic_metadata.mcast_grp",
      ["standard_metadata", "mcast_grp"]
    ],
    [
      "intrinsic_metadata.resubmit_flag",
      ["standard_metadata", "resubmit_flag"]
    ],
    [
      "intrinsic_metadata.egress_rid",
      ["standard_metadata", "egress_rid"]
    ]
  ]
}