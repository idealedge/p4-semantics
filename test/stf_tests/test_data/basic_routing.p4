/*
Copyright 2013-present Barefoot Networks, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

header_type ethernet_t {
    fields {
        dstAddr : 48;
        srcAddr : 48;
        etherType : 16;
    }
}

header_type ipv4_t {
    fields {
        version : 4;
        ihl : 4;
        diffserv : 8;
        totalLen : 16;
        identification : 16;
        flags : 3;
        fragOffset : 13;
        ttl : 8;
        protocol : 8;
        hdrChecksum : 16;
        srcAddr : 32;
        dstAddr: 32;
    }
}

parser start {
    return parse_ethernet;
}



header ethernet_t ethernet;

parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {
        0x0800 : parse_ipv4;
        default: ingress;
    }
}

header ipv4_t ipv4;

field_list ipv4_checksum_list {
        ipv4.version;
        ipv4.ihl;
        ipv4.diffserv;
        ipv4.totalLen;
        ipv4.identification;
        ipv4.flags;
        ipv4.fragOffset;
        ipv4.ttl;
        ipv4.protocol;
        ipv4.srcAddr;
        ipv4.dstAddr;
}

field_list_calculation ipv4_checksum {
    input {
        ipv4_checksum_list;
    }
    algorithm : csum16;
    output_width : 16;
}

calculated_field ipv4.hdrChecksum {
    verify ipv4_checksum;
    update ipv4_checksum;
}

parser parse_ipv4 {
    extract(ipv4);
    return ingress;
}
header_type ingress_metadata_t {
    fields {
        vrf : 12;
        bd : 16;
        nexthop_index : 16;
    }
}

metadata ingress_metadata_t ingress_metadata;

action on_miss() {
}

action set_bd(bd) {
    modify_field(ingress_metadata.bd, bd);
}

table port_mapping {
    reads {
        standard_metadata.ingress_port : exact;
    }
    actions {
        set_bd;
    }
    size : 32768;
}

action set_vrf(vrf) {
    modify_field(ingress_metadata.vrf, vrf);
}

table bd {
    reads {
        ingress_metadata.bd : exact;
    }
    actions {
        set_vrf;
    }
    size : 65536;
}

action fib_hit_nexthop(nexthop_index) {
    modify_field(ingress_metadata.nexthop_index, nexthop_index);
    subtract_from_field(ipv4.ttl, 1);
}

table ipv4_fib {
    reads {
        ingress_metadata.vrf : exact;
        ipv4.dstAddr : exact;
    }
    actions {
        on_miss;
        fib_hit_nexthop;
    }
    size : 131072;
}

table ipv4_fib_lpm {
    reads {
        ingress_metadata.vrf : exact;
        ipv4.dstAddr : lpm;
    }
    actions {
        on_miss;
        fib_hit_nexthop;
    }
    size : 16384;
}

action set_egress_details(egress_spec) {
    modify_field(standard_metadata.egress_spec, egress_spec);
}

table nexthop {
    reads {
        ingress_metadata.nexthop_index : exact;
    }
    actions {
        on_miss;
        set_egress_details;
    }
    size : 32768;
}

control ingress {
    if (valid(ipv4)) {

        apply(port_mapping);


        apply(bd);


        apply(ipv4_fib) {
            on_miss {
                apply(ipv4_fib_lpm);
            }
        }


        apply(nexthop);
    }
}

action rewrite_src_dst_mac(smac, dmac) {
    modify_field(ethernet.srcAddr, smac);
    modify_field(ethernet.dstAddr, dmac);
}

table rewrite_mac {
    reads {
        ingress_metadata.nexthop_index : exact;
    }
    actions {
        on_miss;
        rewrite_src_dst_mac;
    }
    size : 32768;
}

control egress {

    apply(rewrite_mac);
}
