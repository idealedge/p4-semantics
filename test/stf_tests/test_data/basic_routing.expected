0000000000000011000000110000001100000011000000110000000100000011000000110000001100000011000000110000100000000000010001010000000000000000001011000000000000000000000000000110010100111111000100010110****010101010000000100000010000000110000010000001010000000000000000000000010000000000110010000000000011001010000000000010000000000000000000000000000000000010000001000000011000001000000010100000110000001110000100000001001000010100000101100001100000011010000111000001111 3
0000000000000001000000010000000100000001000000010000000100000001000000010000000100000001000000010000100000000000010001010000000000000000001011000000000000000000000000000110010100111111000100011011****101011000000000100000010000000110000010011000000101010000000000100000010000000000110010000000000011001010000000000010000000000000000000000000000000000010000001000000011000001000000010100000110000001110000100000001001000010100000101100001100000011010000111000001111 1